From 65a72dbc9934bfa5affe6392e995e8b8d3c63f61 Mon Sep 17 00:00:00 2001
From: Hanlu Li <heuleehanlu@gmail.com>
Date: Sat, 19 Apr 2025 15:12:44 +0800
Subject: [PATCH 1052/1058] LATX, fix: Add support for AOSC OS in syscall and
 meson configuration

Handling of 'reserved' fields in struct v4l2_ext_controls differently.

Signed-off-by: Hanlu Li <heuleehanlu@gmail.com>
---
 linux-user/syscall.c | 36 +++++++++++++-----------------------
 meson.build          | 24 ++----------------------
 2 files changed, 15 insertions(+), 45 deletions(-)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 5284c1f2ed..fd16948b05 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -121,19 +121,17 @@
 #ifdef HAVE_BTRFS_H
 #include <linux/btrfs.h>
 #endif
-#ifdef HAVE_DRM_H
-#if defined(IS_LOONGNIX) || defined(IS_ARCH)
+#if defined(HAVE_LIBDRM_H)
 #include <libdrm/drm.h>
 #include <libdrm/i915_drm.h>
 #include <libdrm/radeon_drm.h>
 #include <libdrm/amdgpu_drm.h>
-#else
+#elif defined(HAVE_DRM_H)
 #include <drm/drm.h>
 #include <drm/i915_drm.h>
 #include <drm/radeon_drm.h>
 #include <drm/amdgpu_drm.h>
 #endif
-#endif
 #include <linux/can/raw.h>
 #include <linux/videodev2.h>
 #include <sound/asequencer.h>
@@ -6487,20 +6485,16 @@ static inline abi_long target_to_host_ext_ctrls(struct v4l2_ext_controls *host_v
 {
     memset(host_ver, 0, sizeof(*host_ver));
     uint64_t controls;
+    size_t error_idx_offset = offsetof(struct v4l2_ext_controls, error_idx);
+    size_t request_fd_offset = error_idx_offset + sizeof(uint32_t);
+    size_t reserved_offset = request_fd_offset + sizeof(uint32_t);
     __get_user(host_ver->ctrl_class, &target_ver->ctrl_class);
     __get_user(host_ver->count, &target_ver->count);
     __get_user(host_ver->error_idx, &target_ver->error_idx);
-#if defined(IS_KYLIN) || defined(IS_NFS)
-    __get_user(host_ver->request_fd,
-                                    &target_ver->reserved[0]);
-    __get_user(host_ver->reserved[0],
-                                    &target_ver->reserved[1]);
-#else
-    __get_user(host_ver->reserved[0],
-                                    &target_ver->reserved[0]);
-    __get_user(host_ver->reserved[1],
+    __get_user(*(int32_t *)((uint64_t)host_ver + request_fd_offset),
+                &target_ver->reserved[0]);
+    __get_user(*(uint32_t *)((uint64_t)host_ver + reserved_offset),
                                     &target_ver->reserved[1]);
-#endif
     __get_user(controls, &target_ver->controls);
     host_ver->controls = (struct v4l2_ext_control *)controls;
     return 0;
@@ -6511,20 +6505,16 @@ static inline void host_to_target_ext_ctrls(
                   struct v4l2_ext_controls *host_ver)
 {
     uint64_t controls = (uint64_t)host_ver->controls;
+    size_t error_idx_offset = offsetof(struct v4l2_ext_controls, error_idx);
+    size_t request_fd_offset = error_idx_offset + sizeof(uint32_t);
+    size_t reserved_offset = request_fd_offset + sizeof(uint32_t);
     __put_user(host_ver->ctrl_class, &target_ver->ctrl_class);
     __put_user(host_ver->count, &target_ver->count);
     __put_user(host_ver->error_idx, &target_ver->error_idx);
-#if defined(IS_KYLIN) || defined(IS_NFS)
-    __put_user(host_ver->request_fd,
-                                    &target_ver->reserved[0]);
-    __put_user(host_ver->reserved[0],
-                                    &target_ver->reserved[1]);
-#else
-    __put_user(host_ver->reserved[0],
+    __put_user(*(int32_t *)((uint64_t)host_ver + request_fd_offset),
                                     &target_ver->reserved[0]);
-    __put_user(host_ver->reserved[1],
+    __put_user(*(uint32_t *)((uint64_t)host_ver + reserved_offset),
                                     &target_ver->reserved[1]);
-#endif
     __put_user(controls, &target_ver->controls);
 }
 
diff --git a/meson.build b/meson.build
index a30002f0ec..6e212407ca 100644
--- a/meson.build
+++ b/meson.build
@@ -203,28 +203,8 @@ config_host_data.set('CONFIG_LIBATTR', have_old_libattr)
 config_host_data.set('CONFIG_GETTID', has_gettid)
 config_host_data.set('CONFIG_MALLOC_TRIM', has_malloc_trim)
 
-lsb_release = find_program('lsb_release', required: false)
-if lsb_release.found()
-  lsb_release_output = run_command('lsb_release', '-si').stdout().strip()
-  if lsb_release_output == 'Loongnix'
-    config_host_data.set('HAVE_DRM_H', cc.has_header('libdrm/drm.h'))
-    config_host_data.set('IS_LOONGNIX', 'true')
-  elif lsb_release_output == 'Arch'
-    config_host_data.set('HAVE_DRM_H', cc.has_header('libdrm/drm.h'))
-    config_host_data.set('IS_ARCH', 'true')
-  elif lsb_release_output == 'Kylin'
-    config_host_data.set('HAVE_DRM_H', cc.has_header('drm/drm.h'))
-    config_host_data.set('IS_KYLIN', 'true')
-  elif lsb_release_output == 'Nfsdesktop'
-    #方德
-    config_host_data.set('HAVE_DRM_H', cc.has_header('drm/drm.h'))
-    config_host_data.set('IS_NFS', 'true')
-  else
-    config_host_data.set('HAVE_DRM_H', cc.has_header('drm/drm.h'))
-  endif
-else
-    config_host_data.set('HAVE_DRM_H', cc.has_header('drm/drm.h'))
-endif
+config_host_data.set('HAVE_LIBDRM_H', cc.has_header('libdrm/drm.h'))
+config_host_data.set('HAVE_DRM_H', cc.has_header('drm/drm.h'))
 
 config_host_data.set('CONFIG_PREADV', cc.has_function('preadv', prefix: '#include <sys/uio.h>'))
 
-- 
2.49.0

